File: engines/scoring.py
"""AI Skor Sistemi — her motorun ürettiği ham skoru (-1..+1) ağırlıklandırır.

Çıktı:
  long_score  0-100
  short_score 100 - long_score
  decision    STRONG LONG / LONG / WEAK LONG / NÖTR / WEAK SHORT / SHORT / STRONG SHORT
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from core.indicators import clamp, tr_lower
from engines import trend as trend_engine

# Gösterge adlarının kullanıcıya gösterilecek karşılıkları
LABELS = {
    "trend": "Trend",
    "open_interest": "Open Interest",
    "funding": "Funding",
    "cvd": "CVD / Order Flow",
    "rsi": "RSI",
    "macd": "MACD",
    "orderbook": "Order Book",
    "whale": "Whale",
    "liquidation": "Likidasyon",
    "smart_money": "Smart Money",
}


def _get(d: Optional[Dict], *path, default: float = 0.0) -> float:
    node: Any = d
    for p in path:
        if not isinstance(node, dict) or p not in node:
            return default
        node = node[p]
    try:
        return float(node)
    except (TypeError, ValueError):
        return default


def collect_components(engines: Dict[str, Any]) -> Dict[str, Dict[str, Any]]:
    """Her bileşen için ham skor (-1..+1) ve kısa gerekçe üretir."""
    trend = engines.get("trend", {})
    deriv = engines.get("derivatives", {})
    flow = engines.get("order_flow", {})
    book = engines.get("orderbook", {})
    whale = engines.get("whale", {})
    sm = engines.get("smart_money", {})

    rsi_part = trend_engine.rsi_score(trend)
    macd_part = trend_engine.macd_score(trend)

    # OI bileşeni, büyük hesap Long/Short oranıyla harmanlanır
    oi_raw = _get(deriv, "open_interest", "score")
    ls_raw = _get(deriv, "long_short", "score")
    oi_component = clamp(oi_raw * 0.7 + ls_raw * 0.3)

    # CVD bileşeni, Binance taker buy/sell dengesiyle harmanlanır
    flow_raw = _get(flow, "score")
    taker_raw = _get(deriv, "taker", "score")
    cvd_component = clamp(flow_raw * 0.7 + taker_raw * 0.3)

    comps: Dict[str, Dict[str, Any]] = {
        "trend": {
            "raw": clamp(_get(trend, "score")),
            "detail": trend.get("summary", ""),
            "available": bool(trend.get("timeframes")),
        },
        "open_interest": {
            "raw": oi_component,
            "detail": _oi_detail(deriv),
            "available": "error" not in deriv.get("open_interest", {}),
        },
        "funding": {
            "raw": clamp(_get(deriv, "funding", "score")),
            "detail": _funding_detail(deriv),
            "available": "error" not in deriv.get("funding", {}),
        },
        "cvd": {
            "raw": cvd_component,
            "detail": flow.get("summary", ""),
            "available": bool(flow.get("available")),
        },
        "rsi": {
            "raw": clamp(rsi_part["score"]),
            "detail": f"RSI {rsi_part.get('value')} — {rsi_part.get('state')}",
            "available": rsi_part.get("value") is not None,
        },
        "macd": {
            "raw": clamp(macd_part["score"]),
            "detail": f"MACD {macd_part.get('state')}",
            "available": macd_part.get("state") != "NA",
        },
        "orderbook": {
            "raw": clamp(_get(book, "score")),
            "detail": book.get("summary", ""),
            "available": bool(book.get("available")),
        },
        "whale": {
            "raw": clamp(_get(whale, "score")),
            "detail": whale.get("summary", ""),
            "available": bool(whale.get("available")),
        },
        "liquidation": {
            "raw": clamp(_get(deriv, "liquidations", "score")),
            "detail": _liq_detail(deriv),
            "available": bool(deriv.get("liquidations", {}).get("available")),
        },
        "smart_money": {
            "raw": clamp(_get(sm, "score")),
            "detail": sm.get("summary", ""),
            "available": bool(sm.get("available")),
        },
    }
    return comps


def _oi_detail(deriv: Dict) -> str:
    oi = deriv.get("open_interest", {})
    if "error" in oi:
        return "Veri yok"
    ch = oi.get("change_1h_pct")
    interp = oi.get("interpretation_1h", {})
    ls = deriv.get("long_short", {})
    txt = f"OI 1s %{ch:+.2f} — {interp.get('state', '')}" if ch is not None else "OI verisi yok"
    if ls.get("top_positions_ratio"):
        txt += f" | Büyük hesap L/S {ls['top_positions_ratio']:.2f}"
    return txt


def _funding_detail(deriv: Dict) -> str:
    f = deriv.get("funding", {})
    if "error" in f or f.get("current_pct") is None:
        return "Veri yok"
    return (f"%{f['current_pct']:+.4f} ({f['health']}, yıllık ~%{f['annualized_pct']:.1f}) "
            f"· {f['bias']}")


def _liq_detail(deriv: Dict) -> str:
    liq = deriv.get("liquidations", {})
    if not liq.get("available"):
        return liq.get("note", "Likidasyon verisi yok")
    return (f"Long {liq['long_usdt']:,.0f} / Short {liq['short_usdt']:,.0f} USDT "
            f"({liq['hours']}s)" + (f" · {liq['squeeze']}" if liq["squeeze"] != "NONE" else ""))


def decide(long_score: float, thresholds: Dict[str, float]) -> str:
    if long_score >= thresholds.get("strong_long", 75):
        return "STRONG LONG"
    if long_score >= thresholds.get("long", 62):
        return "LONG"
    if long_score >= thresholds.get("weak_long", 55):
        return "WEAK LONG"
    if long_score <= thresholds.get("strong_short", 25):
        return "STRONG SHORT"
    if long_score <= thresholds.get("short", 38):
        return "SHORT"
    if long_score <= thresholds.get("weak_short", 45):
        return "WEAK SHORT"
    return "NÖTR"


def run(engines: Dict[str, Any], cfg) -> Dict[str, Any]:
    weights: Dict[str, float] = dict(cfg.get("scoring.weights", {}))
    thresholds: Dict[str, float] = dict(cfg.get("scoring.thresholds", {}))
    comps = collect_components(engines)

    rows: List[Dict[str, Any]] = []
    total_weight = 0.0
    weighted_sum = 0.0
    available_weight = 0.0

    for key, w in weights.items():
        c = comps.get(key)
        if c is None:
            continue
        raw = c["raw"] if c["available"] else 0.0
        points = raw * w
        weighted_sum += points
        total_weight += w
        if c["available"]:
            available_weight += w
        rows.append({
            "key": key,
            "name": LABELS.get(key, key),
            "weight": w,
            "raw": round(raw, 4),
            "points": round(points, 2),
            "max_points": w,
            "available": c["available"],
            "detail": c["detail"],
            "direction": ("LONG" if raw > 0.05 else "SHORT" if raw < -0.05 else "NÖTR"),
        })

    net = weighted_sum / total_weight if total_weight else 0.0     # -1 .. +1
    long_score = round(50 + 50 * net, 1)
    long_score = max(0.0, min(100.0, long_score))
    short_score = round(100 - long_score, 1)
    decision = decide(long_score, thresholds)

    # Güven: veri kapsaması + bileşenler arası uyum
    coverage = available_weight / total_weight if total_weight else 0.0
    directions = [r["raw"] for r in rows if r["available"] and abs(r["raw"]) > 0.05]
    if directions:
        agree = sum(1 for d in directions if (d > 0) == (net > 0)) / len(directions)
    else:
        agree = 0.5
    confidence = round(clamp(coverage * 0.5 + agree * 0.5, 0, 1) * 100, 1)

    rows_sorted = sorted(rows, key=lambda r: abs(r["points"]), reverse=True)

    return {
        "long_score": long_score,
        "short_score": short_score,
        "net": round(net, 4),
        "decision": decision,
        "confidence": confidence,
        "coverage_pct": round(coverage * 100, 1),
        "agreement_pct": round(agree * 100, 1),
        "components": rows,
        "top_drivers": rows_sorted[:4],
        "total_points": round(weighted_sum, 2),
        "total_weight": total_weight,
        "answer": _answer(engines, decision, long_score),
    }


def _answer(engines: Dict[str, Any], decision: str, long_score: float) -> str:
    """'Büyük para şu anda ne yapıyor?' sorusuna tek cümlelik cevap."""
    deriv = engines.get("derivatives", {})
    whale = engines.get("whale", {})
    flow = engines.get("order_flow", {})

    oi_state = deriv.get("open_interest", {}).get("interpretation_1h", {}).get("state", "")
    whale_state = whale.get("state", "")
    div = flow.get("divergence_note", "")
    ls = deriv.get("long_short", {}).get("top_positions_ratio")

    bits: List[str] = []
    if oi_state and oi_state != "NA":
        bits.append(f"pozisyonlanma: {tr_lower(oi_state)}")
    if ls:
        taraf = "long" if ls > 1 else "short"
        bits.append(f"büyük hesaplar {taraf} ağırlıklı (L/S {ls:.2f})")
    if whale_state and whale_state != "NÖTR":
        bits.append(f"balinalar {tr_lower(whale_state)} yapıyor")
    if div:
        bits.append(tr_lower(div))

    core = "; ".join(bits) if bits else "belirgin bir yön yok"
    return f"{decision} (Long {long_score:.0f}/100) — {core}."

File: docs/js/scoring.js
// AI Skor Sistemi · Risk Engine · Trade Setup Engine
// Python sürümündeki engines/scoring.py, risk.py, trade_setup.py karşılığı.

import { macdScore, rsiScore } from './engines.js';
import { clamp, trLower } from './util.js';

const LABELS = {
  trend: 'Trend',
  open_interest: 'Open Interest',
  funding: 'Funding',
  cvd: 'CVD / Order Flow',
  rsi: 'RSI',
  macd: 'MACD',
  orderbook: 'Order Book',
  whale: 'Whale',
  liquidation: 'Likidasyon',
  smart_money: 'Smart Money',
};

function oiDetail(d) {
  const oi = d.openInterest || {};
  if (oi.change1hPct == null) return 'OI verisi yok';
  let txt = `OI 1s %${oi.change1hPct.toFixed(2)} — ${oi.interpretation1h?.state || ''}`;
  if (d.longShort?.topPositionsRatio) txt += ` | Büyük hesap L/S ${d.longShort.topPositionsRatio.toFixed(2)}`;
  return txt;
}

function fundingDetail(d) {
  const f = d.funding || {};
  if (f.currentPct == null) return 'Veri yok';
  return `%${f.currentPct.toFixed(4)} (${f.health}, yıllık ~%${f.annualizedPct.toFixed(1)}) · ${f.bias}`;
}

export function collectComponents(e) {
  const { trend, derivatives: d, orderFlow: flow, orderbook: book, whale, smartMoney: sm } = e;
  const rsiPart = rsiScore(trend);
  const macdPart = macdScore(trend);

  const oiComponent = clamp((d.openInterest?.score ?? 0) * 0.7 + (d.longShort?.score ?? 0) * 0.3);
  const cvdComponent = clamp((flow.score ?? 0) * 0.7 + (d.taker?.score ?? 0) * 0.3);

  return {
    trend: { raw: clamp(trend.score), detail: trend.summary, available: !!trend.timeframes },
    open_interest: { raw: oiComponent, detail: oiDetail(d), available: d.openInterest?.change1hPct != null },
    funding: { raw: clamp(d.funding?.score ?? 0), detail: fundingDetail(d), available: d.funding?.currentPct != null },
    cvd: { raw: cvdComponent, detail: flow.summary, available: !!flow.available },
    rsi: { raw: clamp(rsiPart.score), detail: `RSI ${rsiPart.value?.toFixed(1)} — ${rsiPart.state}`, available: rsiPart.value != null },
    macd: { raw: clamp(macdPart.score), detail: `MACD ${macdPart.state}`, available: macdPart.state !== 'NA' },
    orderbook: { raw: clamp(book.score ?? 0), detail: book.summary, available: !!book.available },
    whale: { raw: clamp(whale.score ?? 0), detail: whale.summary, available: !!whale.available },
    liquidation: { raw: 0, detail: d.liquidations?.note || 'Veri yok', available: false },
    smart_money: { raw: clamp(sm.score ?? 0), detail: sm.summary, available: !!sm.available },
  };
}

function decide(longScore, t) {
  if (longScore >= t.strongLong) return 'STRONG LONG';
  if (longScore >= t.long) return 'LONG';
  if (longScore >= t.weakLong) return 'WEAK LONG';
  if (longScore <= t.strongShort) return 'STRONG SHORT';
  if (longScore <= t.short) return 'SHORT';
  if (longScore <= t.weakShort) return 'WEAK SHORT';
  return 'NÖTR';
}

function answer(e, decision, longScore) {
  const oiState = e.derivatives.openInterest?.interpretation1h?.state;
  const ls = e.derivatives.longShort?.topPositionsRatio;
  const bits = [];
  if (oiState && oiState !== 'NA') bits.push(`pozisyonlanma: ${trLower(oiState)}`);
  if (ls) bits.push(`büyük hesaplar ${ls > 1 ? 'long' : 'short'} ağırlıklı (L/S ${ls.toFixed(2)})`);
  if (e.whale.state && e.whale.state !== 'NÖTR') bits.push(`balinalar ${trLower(e.whale.state)} yapıyor`);
  if (e.orderFlow.divergenceNote) bits.push(trLower(e.orderFlow.divergenceNote));
  return `${decision} (Long ${longScore.toFixed(0)}/100) — ${bits.join('; ') || 'belirgin bir yön yok'}.`;
}

export function scoreEngine(engines, cfg) {
  const weights = cfg.scoring.weights;
  const comps = collectComponents(engines);
  const rows = [];
  let totalWeight = 0, weightedSum = 0, availableWeight = 0;

  for (const [key, w] of Object.entries(weights)) {
    const c = comps[key];
    if (!c) continue;
    const raw = c.available ? c.raw : 0;
    const points = raw * w;
    weightedSum += points;
    totalWeight += w;
    if (c.available) availableWeight += w;
    rows.push({
      key, name: LABELS[key] || key, weight: w, raw, points, maxPoints: w,
      available: c.available, detail: c.detail,
      direction: raw > 0.05 ? 'LONG' : raw < -0.05 ? 'SHORT' : 'NÖTR',
    });
  }

  const net = totalWeight ? weightedSum / totalWeight : 0;
  const longScore = Math.max(0, Math.min(100, 50 + 50 * net));
  const shortScore = 100 - longScore;
  const decision = decide(longScore, cfg.scoring.thresholds);

  const coverage = totalWeight ? availableWeight / totalWeight : 0;
  const dirs = rows.filter((r) => r.available && Math.abs(r.raw) > 0.05).map((r) => r.raw);
  const agree = dirs.length ? dirs.filter((d) => d > 0 === net > 0).length / dirs.length : 0.5;
  const confidence = clamp(coverage * 0.5 + agree * 0.5, 0, 1) * 100;

  return {
    longScore, shortScore, net, decision, confidence,
    coveragePct: coverage * 100, agreementPct: agree * 100,
    components: rows,
    topDrivers: rows.slice().sort((a, b) => Math.abs(b.points) - Math.abs(a.points)).slice(0, 4),
    totalPoints: weightedSum, totalWeight,
    answer: answer(engines, decision, longScore),
  };
}

// ===========================================================================
// RISK ENGINE
// ===========================================================================
export function riskEngine(engines, cfg) {
  const r = cfg.risk;
  const { trend, derivatives: d, orderbook: book } = engines;
  const factors = [];
  let points = 0;

  const atrPct = trend.atrPct;
  if (atrPct != null) {
    const hv = r.highVolatilityAtrPct;
    let p, state;
    if (atrPct >= hv * 1.5) { p = 3; state = 'ÇOK YÜKSEK'; }
    else if (atrPct >= hv) { p = 2; state = 'YÜKSEK'; }
    else if (atrPct >= hv * 0.5) { p = 1; state = 'NORMAL'; }
    else { p = 0; state = 'DÜŞÜK'; }
    points += p;
    factors.push({ factor: 'Volatilite (ATR%)', value: `%${atrPct.toFixed(2)}`, state, points: p });
  }

  const cur = d.funding?.currentPct;
  if (cur != null) {
    const ex = r.extremeFunding;
    const a = Math.abs(cur);
    let p, state;
    if (a >= ex * 2) { p = 3; state = 'AŞIRI'; }
    else if (a >= ex) { p = 2; state = 'YÜKSEK'; }
    else if (a >= ex / 2) { p = 1; state = 'ARTIYOR'; }
    else { p = 0; state = 'SAĞLIKLI'; }
    points += p;
    factors.push({ factor: 'Funding maliyeti', value: `%${cur.toFixed(4)}`, state, points: p });
  }

  const ls = d.longShort?.topPositionsRatio;
  if (ls) {
    const c = r.crowdedLsRatio;
    let p, state;
    if (ls >= c || ls <= 1 / c) { p = 2; state = 'KALABALIK'; }
    else if (ls >= c * 0.75 || ls <= 1 / (c * 0.75)) { p = 1; state = 'YOĞUNLAŞIYOR'; }
    else { p = 0; state = 'DENGELİ'; }
    points += p;
    factors.push({ factor: 'Pozisyon kalabalığı', value: `L/S ${ls.toFixed(2)}`, state, points: p });
  }

  if (book?.available) {
    const spread = book.spreadPct;
    let p, state;
    if (spread > 0.08) { p = 2; state = 'GENİŞ SPREAD'; }
    else if (spread > 0.03) { p = 1; state = 'ORTA'; }
    else { p = 0; state = 'DERİN'; }
    points += p;
    factors.push({ factor: 'Order book likiditesi', value: `%${spread.toFixed(4)} spread`, state, points: p });
    if (book.spoofs?.length) {
      points += 1;
      factors.push({ factor: 'Spoof aktivitesi', value: `${book.spoofs.length} şüpheli duvar`, state: 'MANİPÜLASYON ŞÜPHESİ', points: 1 });
    }
  }

  if (trend.timeframes && !trend.mtfAligned) {
    points += 1;
    factors.push({ factor: 'Zaman dilimi uyumu', value: 'Uyumsuz', state: 'KARARSIZ', points: 1 });
  }

  const level = points >= 6 ? 'YÜKSEK' : points >= 3 ? 'ORTA' : 'DÜŞÜK';
  const riskUsdt = (r.accountSizeUsdt * r.riskPerTradePct) / 100;
  const levFactor = { DÜŞÜK: 1.0, ORTA: 0.6, YÜKSEK: 0.3 }[level];

  return {
    level, points, factors,
    accountSizeUsdt: r.accountSizeUsdt, riskPerTradePct: r.riskPerTradePct, riskUsdt,
    suggestedMaxLeverage: Math.max(1, Math.round(r.maxLeverage * levFactor)),
    summary: `Risk ${level} (${points} puan) · ` + factors.slice(0, 3).map((f) => `${f.factor}: ${f.state}`).join(' · '),
  };
}

// ===========================================================================
// TRADE SETUP ENGINE
// ===========================================================================
function roundTick(value, tick, precision) {
  if (!value) return value;
  let v = value;
  if (tick && tick > 0) v = Math.round(v / tick) * tick;
  return +v.toFixed(precision);
}

function nearestZones(sm, direction, price) {
  const out = [];
  const want = direction === 'LONG' ? 'bullish' : 'bearish';
  for (const f of sm.fvg?.open || []) {
    if (f.direction !== want) continue;
    if (direction === 'LONG' && f.top <= price * 1.002) out.push({ kind: 'FVG', low: f.bottom, high: f.top, distancePct: f.distancePct });
    else if (direction === 'SHORT' && f.bottom >= price * 0.998) out.push({ kind: 'FVG', low: f.bottom, high: f.top, distancePct: f.distancePct });
  }
  for (const o of sm.orderBlocks?.fresh || []) {
    if (o.direction !== want) continue;
    if (direction === 'LONG' && o.top <= price * 1.002) out.push({ kind: 'OB', low: o.bottom, high: o.top, distancePct: o.distancePct });
    else if (direction === 'SHORT' && o.bottom >= price * 0.998) out.push({ kind: 'OB', low: o.bottom, high: o.top, distancePct: o.distancePct });
  }
  return out.sort((a, b) => Math.abs(a.distancePct) - Math.abs(b.distancePct)).slice(0, 3);
}

function snapTarget(target, candidates, direction, tolerancePct = 1.2) {
  for (const c of candidates) {
    if (!c) continue;
    if ((Math.abs(c - target) / target) * 100 <= tolerancePct) {
      return c * (direction === 'LONG' ? 0.999 : 1.001);
    }
  }
  return target;
}

export function tradeSetupEngine(symbol, engines, score, risk, cfg, meta) {
  const s = cfg.setup;
  const { trend, smartMoney: sm, derivatives: d, orderbook: book } = engines;
  const price = trend.price ?? book.midPrice;
  const mtf = trend.timeframes?.mtf;
  const atr = mtf?.atr?.value;
  const tick = meta?.tickSize ?? 0;
  const precision = meta?.pricePrecision ?? 8;

  let direction, conviction;
  if (score.longScore >= s.minScoreForSetup) { direction = 'LONG'; conviction = score.longScore; }
  else if (score.shortScore >= s.minScoreForSetup) { direction = 'SHORT'; conviction = score.shortScore; }
  else {
    return {
      available: false, direction: 'NONE', recommendation: 'BEKLE',
      reason: `Skor nötr bölgede (Long ${score.longScore.toFixed(0)}/100). Setup için en az ${s.minScoreForSetup} gerekiyor.`,
    };
  }
  if (!price || !atr || atr <= 0) {
    return { available: false, direction, recommendation: 'BEKLE', reason: 'Fiyat/ATR verisi eksik' };
  }

  const zones = nearestZones(sm, direction, price);
  const pullback = s.entryPullbackAtr * atr;
  const vwapD = mtf.vwap?.daily;
  let entryLow, entryHigh, entryBasis;

  if (zones.length) {
    const z = zones[0];
    entryLow = Math.min(z.low, z.high);
    entryHigh = Math.max(z.low, z.high);
    entryBasis = `${z.kind} bölgesi`;
  } else if (direction === 'LONG') {
    entryHigh = price;
    entryLow = price - pullback;
    if (vwapD && price > vwapD && vwapD > price - 2 * atr) entryLow = Math.min(entryLow, vwapD);
    entryBasis = 'ATR geri çekilmesi' + (vwapD ? ' + Günlük VWAP' : '');
  } else {
    entryLow = price;
    entryHigh = price + pullback;
    if (vwapD && price < vwapD && vwapD < price + 2 * atr) entryHigh = Math.max(entryHigh, vwapD);
    entryBasis = 'ATR geri çekilmesi' + (vwapD ? ' + Günlük VWAP' : '');
  }
  const entry = (entryLow + entryHigh) / 2;

  const swingLow = mtf.structure?.lastSwingLow;
  const swingHigh = mtf.structure?.lastSwingHigh;
  let stop, stopBasis;
  if (direction === 'LONG') {
    const atrStop = entry - s.stopAtrMultiplier * atr;
    const structStop = swingLow && swingLow < entry ? swingLow - 0.25 * atr : null;
    stop = structStop != null ? Math.min(atrStop, structStop) : atrStop;
    stopBasis = structStop != null && structStop <= atrStop ? 'Swing low altı' : `${s.stopAtrMultiplier}x ATR`;
  } else {
    const atrStop = entry + s.stopAtrMultiplier * atr;
    const structStop = swingHigh && swingHigh > entry ? swingHigh + 0.25 * atr : null;
    stop = structStop != null ? Math.max(atrStop, structStop) : atrStop;
    stopBasis = structStop != null && structStop >= atrStop ? 'Swing high üstü' : `${s.stopAtrMultiplier}x ATR`;
  }

  const rDistance = Math.abs(entry - stop);
  if (rDistance <= 0) {
    return { available: false, direction, recommendation: 'BEKLE', reason: 'Geçersiz stop mesafesi' };
  }

  const structural = [];
  if (direction === 'LONG') {
    if (swingHigh) structural.push(swingHigh);
    for (const f of sm.fvg?.open || []) if (f.direction === 'bearish' && f.bottom > entry) structural.push(f.bottom);
    for (const w of book.askWalls || []) if (w.price > entry) structural.push(w.price);
  } else {
    if (swingLow) structural.push(swingLow);
    for (const f of sm.fvg?.open || []) if (f.direction === 'bullish' && f.top < entry) structural.push(f.top);
    for (const w of book.bidWalls || []) if (w.price < entry) structural.push(w.price);
  }

  const targets = s.tpRMultiples.map((m, i) => {
    const raw = direction === 'LONG' ? entry + rDistance * m : entry - rDistance * m;
    const snapped = snapTarget(raw, structural, direction);
    return {
      name: `TP${i + 1}`, price: roundTick(snapped, tick, precision), rMultiple: m,
      gainPct: ((snapped - entry) / entry) * 100 * (direction === 'LONG' ? 1 : -1),
      snapped: Math.abs(snapped - raw) > 1e-12,
    };
  });

  const entryR = roundTick(entry, tick, precision);
  const stopR = roundTick(stop, tick, precision);
  const dist = Math.abs(entryR - stopR);
  const qty = dist ? risk.riskUsdt / dist : 0;

  const invalidation = [];
  if (direction === 'LONG') {
    if (swingLow) invalidation.push(`1H swing low ${roundTick(swingLow, tick, precision)} altında kapanış`);
    invalidation.push("SuperTrend yönünün DOWN'a dönmesi");
    invalidation.push('OI artarken fiyatın düşmesi (yeni short girişi)');
  } else {
    if (swingHigh) invalidation.push(`1H swing high ${roundTick(swingHigh, tick, precision)} üstünde kapanış`);
    invalidation.push("SuperTrend yönünün UP'a dönmesi");
    invalidation.push('OI artarken fiyatın yükselmesi (yeni long girişi)');
  }

  return {
    available: true, symbol, direction, recommendation: score.decision,
    trend: trend.label,
    probability: Math.min(95, 0.65 * conviction + 0.35 * score.confidence),
    confidence: score.confidence, price,
    entry: entryR,
    entryZone: [roundTick(Math.min(entryLow, entryHigh), tick, precision), roundTick(Math.max(entryLow, entryHigh), tick, precision)],
    entryBasis, stop: stopR, stopBasis,
    stopDistancePct: entryR ? (dist / entryR) * 100 : 0,
    targets, riskReward: targets[0]?.rMultiple ?? 0,
    rrOk: (targets[0]?.rMultiple ?? 0) >= s.minRr,
    position: {
      riskUsdt: risk.riskUsdt, qty, notionalUsdt: qty * entryR,
      suggestedMaxLeverage: risk.suggestedMaxLeverage,
    },
    context: {
      funding: d.funding?.health ?? 'NA', fundingPct: d.funding?.currentPct,
      oi: d.openInterest?.trend ?? 'NA', oiChange1h: d.openInterest?.change1hPct,
      cvd: engines.orderFlow.label ?? 'NA', whales: engines.whale.state ?? 'NA',
      risk: risk.level, atrPct: mtf.atr?.pct,
    },
    invalidation,
  };
}
